internal/pkg/instrumentation/bpf/go.opentelemetry.io/auto/sdk/bpf/probe.bpf.c

```diff
@@ -0,0 +1,219 @@
+	}
+
+	return rc;
+}
+
+SEC("uprobe/Tracer_start")
+int uprobe_Tracer_start(struct pt_regs *ctx) {
+    struct go_iface go_context = {0};
+    get_Go_context(ctx, 2, 0, true, &go_context);
+
+	struct otel_span_t otel_span;
+    __builtin_memset(&otel_span, 0, sizeof(struct otel_span_t));
+
+    start_span_params_t params = {
+        .ctx = ctx,
+        .go_context = &go_context,
+        .psc = &otel_span.psc,
+        .sc = &otel_span.sc,
+        .get_parent_span_context_fn = NULL,
+        .get_parent_span_context_arg = NULL, // Default to new root. 
+    };
+
+    start_span(&params);
+
+	void *parent_span_context = get_argument(ctx, 5);
+	long rc = write_span_context(parent_span_context, &otel_span.psc);
+	if (rc != 0) {
+		bpf_printk("failed to write parent span context: %ld", rc);
+	}
+
+	if (!is_sampled(params.sc)) {
+		// Default SDK behaviour is to sample everything. Only set the sampled
+		// value to false when needed.
+		void *sampled_ptr_val = get_argument(ctx, 6);
+		if (sampled_ptr_val == NULL) {
+			bpf_printk("nil sampled pointer");
+		} else {
+			bool false_val = false;
+			rc = bpf_probe_write_user(sampled_ptr_val, &false_val, sizeof(bool));
+			if (rc != 0) {
+				bpf_printk("bpf_probe_write_user: failed to write sampled value: %ld", rc);
+			} else {
+				bpf_printk("wrote sampled value");
+			}
+		}
+	}
+
+	void *span_context_ptr_val = get_argument(ctx, 7);
+	rc = write_span_context(span_context_ptr_val, &otel_span.sc);
+	if (rc != 0) {
+		bpf_printk("failed to write span context: %ld", rc);
+	}
+
+	void *span_ptr_val = get_argument(ctx, 4);
+    bpf_map_update_elem(&active_spans_by_span_ptr, &span_ptr_val, &otel_span, 0);
+    start_tracking_span(go_context.data, &otel_span.sc);
+
+	return 0;
+}
+
+SEC("uprobe/Span_ended")
+int uprobe_Span_ended(struct pt_regs *ctx) {
+    void *span_ptr = get_argument(ctx, 1);
+    struct otel_span_t *span = bpf_map_lookup_elem(&active_spans_by_span_ptr, &span_ptr);
+    if (span == NULL) {
+        return 0;
+    }
+    stop_tracking_span(&span->sc, &span->psc);
+    bpf_map_delete_elem(&active_spans_by_span_ptr, &span_ptr);
+
+	u64 len = (u64)get_argument(ctx, 3);
+	if (len > MAX_SIZE) {
+		bpf_printk("span data too large: %d", len);
+		return -1;
+	}
+	if (len == 0) {
+		bpf_printk("empty span data");
+		return 0;
+	}
+
+	void *data_ptr = get_argument(ctx, 2);
+	if (data_ptr == NULL) {
+		bpf_printk("empty span data");
+		return 0;
+	}
+
+	u32 key = 0;
+	struct event_t *event = bpf_map_lookup_elem(&new_event, &key);
+	if (event == NULL) {
+		bpf_printk("failed to initialize new event");
+		return -2;
+	}
+	event->size = (u32)len;
+
+	if (event->size < MAX_SIZE) {
+		long rc = bpf_probe_read(&event->data, event->size, data_ptr);
+		if (rc < 0) {
+			bpf_printk("failed to read encoded span data");
+			return -3;
+		}
+	} else {
+		bpf_printk("read too large: %d", event->size);
+		return -4;
+	}
+
+	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU, event, sizeof(*event));
```

Comment:
We can also save space on the perf buffer in this case, since we know the actual encoding size we don't have to output all the event buffer.

Reply:
How would we do that? My understanding is that the event buffer (i.e. `event.data`) is allocated to the stack? How do we dynamically send a smaller buffer?

Would the Go `event` definition also need to be updated? Would it still be an array?

Reply:
How do we determine the size though? Is it as simple as `sizeof(event.size)` (i.e. 4) + `event.size`?

Reply:
I thinks so

Reply:
Flattening `output_span_event` and checking size before `bpf_perf_event_output` got around the last error. However, now the base64 decoding is failing:

```terminal
{
  "time": "2024-09-20T10:22:43.749409627-07:00",
  "level": "ERROR",
  "source": {
    "function": "go.opentelemetry.io/auto/internal/pkg/instrumentation/probe.(*Base[...]).Run",
    "file": "/home/tyler/go/src/go.opentelemetry.io/auto/internal/pkg/instrumentation/probe/probe.go",
    "line": 216
  },
  "msg": "failed to process perf record",
  "error": "unexpected EOF"
}
```

I'm guessing this because it is trying to deserialize into an object larger than the serialized data now that we have shortened.  I'm going to verify with some ad hoc logging, but I expect this change might not be viable.

Reply:
There might be 4 bytes of padding between the size and data fields which causes the parsing error.

Reply:
Hmm, not sure. Reading an extra 4 bytes after the `Size` didn't change the error. This involved writing a custom `perf.Record` parser for the SDK probe.

I was able to add more logging information and those error messages are coming from the `net/http` and `database/sql` probes. It looks like this is corrupting those perf events.

Reply:
Yes, doing this will require having a different parsing of perf events. The generic one assumes constant size events.

Reply:
Gah, looks like I had a staged change that I didn't realize. I think I may have this working, disregard the last comment. :facepalm:

Reply:
Updated. PTAL

Note, to support dynamically outputting the event, we cannot use `output_span_event`. The verifier complains that the size passed is not verified to not exceed the allowed memory. I have flattened it into the function.